cloud/services/domains.go

```diff
@@ -0,0 +1,200 @@
+package services
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/linode/linodego"
+	kutil "sigs.k8s.io/cluster-api/util"
+
+	"github.com/linode/cluster-api-provider-linode/cloud/scope"
+)
+
+var dnsTTLSec = 30
+
+// AddIPToDNS creates domain record for machine public ip
+func AddIPToDNS(
+	ctx context.Context,
+	logger logr.Logger,
+	machineScope *scope.MachineScope,
+) error {
+	// Check if instance is a control plane node
+	if !kutil.IsControlPlaneMachine(machineScope.Machine) {
+		return nil
+	}
+
+	// Get the public IP that was assigned
+	publicIP, err := GetMachinePublicIP(ctx, logger, machineScope)
+	if err != nil {
+		logger.Error(err, "Failed to get public IP of machine")
+		return err
+	}
+
+	// Get domainID from domain name
+	domainID, err := GetDomainID(ctx, logger, machineScope)
+	if err != nil {
+		logger.Error(err, "Failed to get domain ID")
+		return err
+	}
+
+	// Check if record exists for this IP and name combo
+	domainHostname := machineScope.LinodeCluster.ObjectMeta.Name + "-" + machineScope.LinodeCluster.Spec.Network.DNSUniqueIdentifier
+	filter, err := json.Marshal(map[string]interface{}{"name": domainHostname, "target": publicIP})
+	if err != nil {
+		logger.Error(err, "Failed to marshal domain filter")
+		return err
+	}
+
+	domainRecords, err := machineScope.LinodeDomainsClient.ListDomainRecords(ctx, domainID, linodego.NewListOptions(0, string(filter)))
+	if err != nil {
+		logger.Error(err, "unable to get current DNS record from API")
+		return err
+	}
+
+	if machineScope.LinodeCluster.Spec.Network.DNSTTLSec != 0 {
+		dnsTTLSec = machineScope.LinodeCluster.Spec.Network.DNSTTLSec
+	}
+
+	// If record doesnt exist, create it else update it
+	if domainRecords == nil {
+		recordReq := linodego.DomainRecordCreateOptions{
+			Type:   "A",
+			Name:   domainHostname,
+			Target: publicIP,
+			TTLSec: dnsTTLSec,
+		}
+
+		_, err := machineScope.LinodeDomainsClient.CreateDomainRecord(ctx, domainID, recordReq)
+		if err != nil {
+			logger.Error(err, "Failed to create domain record")
+			return err
+		}
+	} else {
+		recordReq := linodego.DomainRecordUpdateOptions{
+			Type:   "A",
+			Name:   domainHostname,
+			Target: publicIP,
+			TTLSec: dnsTTLSec,
+		}
+
+		_, err := machineScope.LinodeDomainsClient.UpdateDomainRecord(ctx, domainID, domainRecords[0].ID, recordReq)
+		if err != nil {
+			logger.Error(err, "Failed to update domain record")
+			return err
+		}
+	}
+
+	return nil
```

Comment:
The nesting can be simplified a bit, since there are only two operations that can occur at this point:

```suggestion
		return nil
	}
	recordReq := linodego.DomainRecordUpdateOptions{
	Type:   "A",
	Name:   domainHostname,
	Target: publicIP,
	TTLSec: dnsTTLSec,
	}

	if _, err := machineScope.LinodeDomainsClient.UpdateDomainRecord(
		ctx,
		domainID,
		domainRecords[0].ID,
		recordReq,
	); err != nil {
		logger.Error(err, "Failed to update domain record")
		return err
	}

	return nil
```

Reply:
updated